src/crystal/scheduler.cr

```diff
@@ -123,4 +138,84 @@ class Crystal::Scheduler
     @current.resume_event.add(0.seconds)
     resume(fiber)
   end
+
+  {% if flag?(:preview_mt) %}
+    @rr_target = 0
+
+    protected def find_target_thread
+      if @@workers.empty?
+        Thread.current
+      else
+        @rr_target += 1
+        @@workers[@rr_target % @@workers.size]
+      end
+
+      # target = Thread.workers[@rr_target]
+      # target_i = @rr_target
+      # (Thread.workers.size - 1).times do |i|
+      #   w_i =(@rr_target + i + 1) % Thread.workers.size
+      #   w = Thread.workers[w_i]
+      #   if w.load < target.load
+      #     target = w
+      #     target_i = w_i
+      #   end
+      # end
+
+      # @rr_target = (target_i + 1) % Thread.workers.size
+      # target
+    end
+
+    def run_loop
+      loop do
+        @lock.lock
+        if runnable = @runnables.shift?
+          @runnables << Fiber.current
+          @lock.unlock
+          runnable.resume
+        else
+          @sleeping = true
+          @lock.unlock
+          oid = @worker_out.read_bytes(UInt64)
+          fiber = Pointer(Fiber).new(oid).as(Fiber)
+          Thread.current.load += 1
+
+          @lock.lock
+          @sleeping = false
+          @runnables << Fiber.current
+          @lock.unlock
+          fiber.resume
+        end
+      end
+    end
+
+    def send_fiber(fiber : Fiber)
+      Thread.current.load -= 1
+      @lock.lock
+      if @sleeping
+        @worker_in.write_bytes(fiber.object_id)
+      else
+        @runnables << fiber
+      end
+      @lock.unlock
+    end
+
+    @@workers = [] of Thread
+
+    def self.workers
+      @@workers
+    end
+
+    def self.init_workers
+      count = ENV["CRYSTAL_WORKERS"]?.try &.to_i || 4
```

Comment:
I dont see this documented anywhere.  should this be added to the docs?
Also should this be an environment variable?  I would expect a command line flag.
What do other languages do to set the number of threads in a program.

Reply:
And why `4`?

Reply:
Perhaps it's the Crystal's equivalent of `GOMAXPROCS`, then it would be better to defaulting it to the number of CPU  cores.

Reply:
Perhaps it's the Crystal equivalent of GOMAXPROCS, then it would be better to defaulting it to the number of CPU cores.

Reply:
For example, you can use GOMAXPROCS in Go to set a similar configuration. Command line flags would require every Crystal app to parse the arguments. Remember this is a dynamic setting: once the program is compiled with MT support (`-D preview_mt`) the number of threads can be changed using that environment variable.

Reply:
Ideally we could use `System.cpu_count` as the default value but we still found high contention that needs to be addressed

Reply:
I actually believe it should be 1 by default and tweakable at compile time (maybe even at runtime).

Defaulting to all available cores only makes sense in some scenarios, for example long lived programs (which can easily define an ENV variable to enable more cores) but may not make sense in short lived programs, where initializing a dozen threads/schedulers will be more costly than just running